Surface_mesh_simplification/include/CGAL/Surface_mesh_simplification/Policies/Edge_collapse/GarlandHeckbert_probabilistic_triangle_policies.h

```diff
@@ -0,0 +1,177 @@
+
+#include <utility>
+
+namespace CGAL {
+namespace Surface_mesh_simplification {
+namespace internal {
+
+template <typename TriangleMesh, typename GeomTraits, typename FaceVarianceMap>
+class Probabilistic_triangle_quadric_calculator
+{
+  typedef typename GeomTraits::FT                                              FT;
+  typedef typename GeomTraits::Point_3                                         Point_3;
+  typedef typename GeomTraits::Vector_3                                        Vector_3;
+
+  typedef typename boost::graph_traits<TriangleMesh>::face_descriptor          face_descriptor;
+
+  typedef Constant_property_map<face_descriptor, FT>                           Default_FVM;
+  typedef typename Default::Get<FaceVarianceMap, Default_FVM>::type            Face_variance_map;
+
+  typedef typename GarlandHeckbert_matrix_types<GeomTraits>::Mat_4             Mat_4;
+  typedef typename GarlandHeckbert_matrix_types<GeomTraits>::Col_4             Col_4;
+
+private:
+  // same meaning as for probabilistic plane quadrics
+  static constexpr FT default_variance_unit = 0.05;
+
+  // this is only used when we fall back to probabilistic planes for the discontinuous edges,
+  // the actual triangle quadric calculation only uses the normal variance
+  static constexpr FT position_variance_factor = 1;
+
+  Face_variance_map m_face_variance_map;
+
+public:
+  Probabilistic_triangle_quadric_calculator() { CGAL_assertion(false); } // compilation only
+
+  template <typename FVM>
+  Probabilistic_triangle_quadric_calculator(const FVM fvm)
+      : m_face_variance_map(fvm)
+  { }
+
+  Probabilistic_triangle_quadric_calculator(TriangleMesh& tmesh,
+                                            typename boost::enable_if<std::is_same<Face_variance_map, Default_FVM> >::type* = nullptr)
+  {
+    // try to initialize the face variance map using the estimated variance
+    // parameters are constants defined for this class
+    FT variance, discard_position;
+    std::tie(variance, discard_position) =
+      estimate_variances(tmesh, GeomTraits(), default_variance_unit, position_variance_factor);
+
+    // see probabilistic plane quadrics
+    m_face_variance_map = Default_FVM { variance };
+  }
+
+public:
+  // we don't have a sensible way to construct a triangle quadric
+  // from an edge, so we fall back to probabilistic plane quadrics here
+  template<typename VertexPointMap>
+  Mat_4 construct_quadric_from_edge(typename boost::graph_traits<TriangleMesh>::halfedge_descriptor he,
+                                    const TriangleMesh& tmesh,
+                                    const VertexPointMap point_map,
+                                    const GeomTraits& gt) const
+  {
+    // same as in probabilistic plane policy
+    const Vector_3 normal = construct_edge_normal(he, tmesh, point_map, gt);
+    const Point_3 p = get(point_map, source(he, tmesh));
+
+    const FT variance = get(m_face_variance_map, face(he, tmesh));
+
+    // @fixme plane?
```

Comment:
is it still a todo? I see there are still other remaining fixme

Reply:
Yes, it is still an active todo